### PR TITLE
[NeoML] switch to VS 2022

### PR DESCRIPTION
**Pull request by favorart**

### CODE REVIEW

**Comment by lordnn** [comment]
Why not to use cmake presets for different compilers?

File: NeoML/src/Dnn/Dnn.cpp [hunk 1]
```diff
@@ -591,23 +591,17 @@ void CDnn::RunOnce()
 		isReuseMemoryMode = ( getOutputBlobsSize() > MinReuseMemoryModeNetSize );
 		runOnce(0);
 	}
-#ifdef NEOML_USE_FINEOBJ
 	catch( CCheckException* exception ) {
 		if( IsLogging() ) {
 			*log << "CCheckException in RunOnce\n";
+#ifdef NEOML_USE_FINEOBJ
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Dnn.cpp:599]: This is wrong. In FineObj exceptions are caught by pointers (compatibility), in open-source - by reference

File: NeoML/test/src/PCATest.cpp [hunk 2]
```diff
@@ -18,6 +18,12 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+//99 - CSVDTest.SVDExampleTest.cpu( Failed )
+//100 - CPCATest.PCAExamplesTest.cpu( Failed )
+//101 - CPCATest.PCAEllipseTest.cpu( Failed )
+//102 - CPCATest.PCASerializationTest.cpu( Failed )
```
Comment (FedyuninV) [inline NeoML/test/src/PCATest.cpp]: Bad comment IMHO. You need only 1 comment for the whole file "these functions are implemented by MKL. MKL is not available on ARM"

File: NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp [hunk 3]
```diff
@@ -450,18 +504,191 @@ void CCpuMathEngine::BlobMaxOverTimePoolingBackward( const CMaxOverTimePoolingDe
...
+				vectorAddValue( sourceRaw, resultRaw, channels, value );
+				sourceRaw += channels;
+				resultRaw += channels;
+			}
 		}
 	}
 }
 
+template<typename T>
+inline void addHeightIndex( const CBlobDesc& source, const CTypedMemoryHandle<const T>& sourceData, bool isForward, const CTypedMemoryHandle<T>& resultData )
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: `addWidthIndex` and `addHeightIndex` are looking very similar. Let's reduce them to one!

File: NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp [hunk 4]
```diff
@@ -450,18 +504,191 @@ void CCpuMathEngine::BlobMaxOverTimePoolingBackward( const CMaxOverTimePoolingDe
...
+			resultRaw += width;
+		}
+	}
+}
+
+//---------------------------------------------------------------------------------------------------
+
+void CCpuMathEngine::AddWidthIndex( const CBlobDesc& source, const CConstFloatHandle& sourceData, bool isForward, const CFloatHandle& resultData )
+{
+	ASSERT_EXPR( sourceData.GetMathEngine() == this );
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: You can pass `*this` to `impl` and move assertions and `scope` to the `impl`.

File: NeoMathEngine/test/src/learn/LUFactorizationTest.cpp [hunk 5]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+#if FINE_ARCHITECTURE( FINE_X86 ) || FINE_ARCHITECTURE( FINE_X64 ) //1 - CMathEngineQRFactorizationTest.Precalc.cpu (Failed) -- DARWIN macOS arm64
```
Comment (FedyuninV) [inline NeoMathEngine/test/src/learn/LUFactorizationTest.cpp]: Same here. It's better to mention the reason of failure (MKL).

File: NeoMathEngine/test/src/learn/MultiplySparseMatrixByMatrixTest.cpp [hunk 6]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+#if FINE_ARCHITECTURE( FINE_X86 ) || FINE_ARCHITECTURE( FINE_X64 ) //44 - CMultiplySparseMatrixByMatrixTestInstantiation.Random.cpu (Failed) -- DARWIN macOS arm64
```
Comment (FedyuninV) [inline NeoMathEngine/test/src/learn/MultiplySparseMatrixByMatrixTest.cpp]: MKL...

File: NeoMathEngine/test/src/learn/MultiplyTransposedMatrixBySparseMatrixTest.cpp [hunk 7]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+#if FINE_ARCHITECTURE( FINE_X86 ) || FINE_ARCHITECTURE( FINE_X64 ) //54 - CMultiplyTransposedMatrixBySparseMatrixTestInstantiation.Random.cpu (Failed) -- DARWIN macOS arm64
```
Comment (FedyuninV) [inline NeoMathEngine/test/src/learn/MultiplyTransposedMatrixBySparseMatrixTest.cpp]: MKL

File: NeoMathEngine/test/src/learn/MultiplyTransposedSparseMatrixByMatrixTest.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+#if FINE_ARCHITECTURE( FINE_X86 ) || FINE_ARCHITECTURE( FINE_X64 ) //29 - CMultiplyTransposedSparseMatrixByMatrixTestInstantiation.Random.cpu (Failed) -- DARWIN macOS arm64
```
Comment (FedyuninV) [inline NeoMathEngine/test/src/learn/MultiplyTransposedSparseMatrixByMatrixTest.cpp]: MKL

File: NeoMathEngine/test/src/learn/QRFactorizationTest.cpp [hunk 9]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 
 #include <TestFixture.h>
 
+#if FINE_ARCHITECTURE( FINE_X86 ) || FINE_ARCHITECTURE( FINE_X64 ) //1 - CMathEngineQRFactorizationTest.Precalc.cpu (Failed) -- DARWIN macOS arm64
```
Comment (FedyuninV) [inline NeoMathEngine/test/src/learn/QRFactorizationTest.cpp]: MKL

**Review by FedyuninV — CHANGES_REQUESTED** [review]

File: NeoMathEngine/test/src/learn/QRFactorizationTest.cpp (hunk 9 above)
Comment (favorart) [reply NeoMathEngine/test/src/learn/QRFactorizationTest.cpp]: fixed

File: NeoMathEngine/test/src/learn/MultiplyTransposedSparseMatrixByMatrixTest.cpp (hunk 8 above)
Comment (favorart) [reply NeoMathEngine/test/src/learn/MultiplyTransposedSparseMatrixByMatrixTest.cpp]: fixed

File: NeoMathEngine/test/src/learn/MultiplyTransposedMatrixBySparseMatrixTest.cpp (hunk 7 above)
Comment (favorart) [reply NeoMathEngine/test/src/learn/MultiplyTransposedMatrixBySparseMatrixTest.cpp]: fixed

File: NeoMathEngine/test/src/learn/MultiplySparseMatrixByMatrixTest.cpp (hunk 6 above)
Comment (favorart) [reply NeoMathEngine/test/src/learn/MultiplySparseMatrixByMatrixTest.cpp]: fixed

File: NeoMathEngine/test/src/learn/LUFactorizationTest.cpp (hunk 5 above)
Comment (favorart) [reply NeoMathEngine/test/src/learn/LUFactorizationTest.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp (hunk 4 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp (hunk 3 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: fixed

File: NeoML/test/src/PCATest.cpp (hunk 2 above)
Comment (favorart) [reply NeoML/test/src/PCATest.cpp]: fixed

File: NeoML/src/Dnn/Dnn.cpp (hunk 1 above)
Comment (favorart) [reply NeoML/src/Dnn/Dnn.cpp:599]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp [hunk 10]
```diff
@@ -450,18 +502,175 @@ void CCpuMathEngine::BlobMaxOverTimePoolingBackward( const CMaxOverTimePoolingDe
...
+	const CBlobDesc& source, const CTypedMemoryHandle<const T>& sourceData, const CTypedMemoryHandle<T>& resultData )
+{
+	ASSERT_EXPR( sourceData.GetMathEngine() == mathEngine );
+	ASSERT_EXPR( resultData.GetMathEngine() == mathEngine );
+	CCpuExecutionScope scope;
+
+	const T* sourceRaw = GetRaw( sourceData );
+	T* resultRaw = GetRaw( resultData );
+
+	if( isWidth ) {
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: You can get rid of thus copy-paste by creating function like `addDimIndex( mathEngine, isForward, objectCount, indexedDimSize, objectSize, sourcePtr, resultPtr )`

Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnPooling.cpp]: Yes, you are right

Final version

```
template<typename T>
static void addDimIndex( const CCpuMathEngine* const mathEngine, bool isForward,
	int objectCount, int indexedDimSize, int operateSize,
	const CTypedMemoryHandle<const T>& sourceData, const CTypedMemoryHandle<T>& resultData )
{
	ASSERT_EXPR( sourceData.GetMathEngine() == mathEngine );
	ASSERT_EXPR( resultData.GetMathEngine() == mathEngine );
	CCpuExecutionScope scope;
	
	const T* sourceRaw = GetRaw( sourceData );
	T* resultRaw = GetRaw( resultData );
	
	for( int batch = 0; batch < objectCount; ++batch ) {
		for( int iDim = 0; iDim < indexedDimSize; ++iDim ) {
			const T value = static_cast<T>( isForward ? iDim : ( -iDim ) );
			vectorAddValue( sourceRaw, resultRaw, operateSize, value );
			sourceRaw += operateSize;
			resultRaw += operateSize;
		}
	}
}
```

**Review by FedyuninV — APPROVED** [review]

**Pull request merged**